src/Common/ZooKeeper/ZooKeeperImpl.cpp

```diff
@@ -309,6 +310,68 @@ ZooKeeper::~ZooKeeper()
     }
 }
 
+namespace
+{
+    using T = std::shared_mutex;
+    using shared_mutex_lock_guard = std::unique_ptr<T, std::function<void(T *)>>;
+
+    void shared_mutex_unlock_exclusive(T * m)
+    {
+        m->unlock();
+    }
+
+    void shared_mutex_unlock_share(T * m)
+    {
+        if (m)
+            m->unlock_shared();
+    }
+
+    auto shared_mutex_lock_exclusive(T * m)
+    {
+        m->lock();
+        return shared_mutex_lock_guard(m, shared_mutex_unlock_exclusive);
+    }
+
+    auto shared_mutex_try_lock_shared(T * m)
+    {
+        if (m->try_lock_shared())
+            return shared_mutex_lock_guard(m, shared_mutex_unlock_share);
+        return shared_mutex_lock_guard(nullptr);
+    }
+}
+
+void ZooKeeper::reloadFaultProbability(const zkutil::ZooKeeperArgs & _args)
```

Comment:
Dynamic reloading looks quite complex. Although, the main usecase is to avoid fault injections during startup and a simple on/off atomic flag would be enough. What do you think about adding a config parameter like `enable_fault_injections_during_startup` and simply checking `Context::getGlobalContextInstance()->isServerCompletelyStarted()` before injecting a fault? It will be useful in our CI as well: https://github.com/ClickHouse/ClickHouse/blob/f540a14fae2a81441113a651357b5bf19904d4f8/docker/test/upgrade/run.sh#L112-L114
JFYI We also have `SYSTEM STOP/START THREAD FUZZER` queries, it's another way.
But dynamic config reloading is okay.